corehq/motech/auth.py

```diff
@@ -39,3 +105,153 @@ def __call__(self, r):
         token = self._get_auth_token(r)
         r.headers["Authorization"] = f"Bearer {token}"
         return r
+
+
+class AuthManager:
+
+    def get_auth(self) -> Optional[AuthBase]:
+        """
+        Returns an instance of requests.auth.AuthBase, to be passed to
+        an outgoing API request, or None if not applicable.
+        """
+        return None
+
+    def get_session(self) -> Session:
+        """
+        Returns an instance of requests.Session. Manages authentication
+        tokens, if applicable.
+        """
+        session = Session()
+        session.auth = self.get_auth()
+        return session
+
+
+class BasicAuthManager(AuthManager):
+
+    def __init__(self, username, password):
+        self.username = username
+        self.password = password
+
+    def get_auth(self):
+        return HTTPBasicAuth(self.username, self.password)
+
+
+class DigestAuthManager(AuthManager):
+
+    def __init__(self, username, password):
+        self.username = username
+        self.password = password
+
+    def get_auth(self):
+        return HTTPDigestAuth(self.username, self.password)
+
+
+class OAuth1Manager(AuthManager):
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        api_endpoints: OAuth1ApiEndpoints,
+        last_token: Optional[dict] = None,
+    ):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.api_endpoints = api_endpoints
+        self.last_token = last_token
+
+    def get_auth(self):
+        assert self.last_token, \
+            'OAuth1 authentication workflow has not been followed for client'
```

Comment:
nit: if this is intended to be a hard guard against this condition occurring in production (rather than a tool for debugging during development), you should raise explicitly, as `assert`s can be dropped (though I'm pretty sure we don't ever use that optimization)